keps/sig-cli/0015-kubectl-plugins.md

```diff
@@ -0,0 +1,152 @@
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - "https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cli/kubectl-extension.md"
+  - "https://github.com/kubernetes/community/pull/481"
+superseded-by
+  - n/a
+---
+
+# Kubectl Plugins 
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Scenarios]](#scenarios)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+This proposal introduces the main design for a plugin mechanism in `kubectl`.
+The mechanism is a git-style system, that looks for executables on a user's `$PATH` whose name begins with `kubectl-`.
+This design allows plugin binaries to override specific command paths, replacing allowed subcommands in existing commands,
+or adding new subcommands under the `kubectl plugin` path.
+
+## Motivation
+
+The existing alpha plugin system in `kubectl` presents a few limitations with its current design. 
+It forces plugin scripts and executables to exist in a pre-determined location, requires basic information for a single 
+plugin to be declared explicitly in its own file, anddoes not provide a clear way to override existing command paths, 
+or provide additional subcommands without having to override a top-level command.
+
+A re-design plugins system allows us to implement extensibility requests from users that the current system cannot address
+(See https://github.com/kubernetes/kubernetes/issues/53640).
+
+### Goals
+
+* Support a simple plugin installation process (no additional config, users drop a binary in their `PATH`, for example, 
+  and they are then able to use them).
+  No additional configuration is needed, only the plugin executable.
+  A plugin's filename determines the plugin's intention, such as which path in the command tree it applies to:
+  `/usr/bin/kubectl-create-foobar` would, for example, override (or create) the command `kubectl create foobar`.
+  It is up to a plugin to parse any arguments and flags given to it. A plugin decides when an argument is a
+  subcommand, as well as any limitations or constraints that its flags should have.
+* Relay all information given to `kubectl` (via command line args) to plugins as-is.
+  Plugins receive all arguments and flags provided by users and are responsible for adjusting their behavior
+  accordingly.
+* Provide a way for plugins to override specific subcommands, while preventing other paths from being overriden.
+* Provide a standard package of common cli utilities that is consumed by `kubectl` and plugins alike.
```

Comment:
I don't think that this goal is related to the plugin mechanism itself.

Reply:
Thanks, will remove this line

Reply:
It's reasonable to move that to non-goal. It's a nice-to-have thing that we will be working on.